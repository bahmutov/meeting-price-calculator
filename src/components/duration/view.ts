import xs from 'xstream';
import * as moment from 'moment';

import { VNode, div, span, button } from '@cycle/dom';
import { State } from './index';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ startTime, duration }) =>
    div('.Duration', [
      div('.start-time', [
        span(`Start time: ${moment(startTime).format('HH:mm:ss')}`),
        button('.reset-button', 'Reset')
      ]),
      div(
        '.duration',
        `Duration: ${moment.duration(duration, 'seconds').humanize()}`
      )
    ])
  );
}
