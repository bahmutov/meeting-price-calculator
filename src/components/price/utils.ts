export default function calculatePrice(
  personAmount: number,
  avgPrice: number,
  tick: number
): number {
  const hourlyToTickRatio: number = 1 / 60 / 60;
  return personAmount * avgPrice * hourlyToTickRatio * tick;
}
