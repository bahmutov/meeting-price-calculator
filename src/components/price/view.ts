import xs from 'xstream';
import { VNode, div, select, option, span, label } from '@cycle/dom';
import { State } from '../../interfaces';
import calculatePrice, { formatPrice } from '../../utils/priceUtils';

function renderPrice(
  personAmountValue: number,
  avgPriceValue: number,
  tick: number,
  currency: string
): VNode {
  const price: number = calculatePrice(personAmountValue, avgPriceValue, tick);
  return div('.Price-actual', [
    div('.Price-label', 'This meeting has cost'),
    div('.Price-value', formatPrice(price, currency))
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      (
        [
          { tick, personAmount, avgPrice, currency },
          personAmountVDom,
          avgPriceVDom
        ]
      ) =>
        div('.Price', [
          renderPrice(personAmount.value, avgPrice.value, tick, currency),
          div('.PriceInputs', [
            personAmountVDom,
            div('.price-result', [
              label('.total-price-label', 'Total price per hour'),
              div(
                '.total-price-value',
                formatPrice(personAmount.value * avgPrice.value, currency)
              ),
              div('.currency', [
                span('.currency-label.label', 'Currency'),
                select('.currency-select', [option('â‚¬'), option('$')])
              ])
            ]),
            avgPriceVDom
          ])
        ])
    );
}
