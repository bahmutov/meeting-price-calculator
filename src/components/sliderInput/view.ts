import xs from 'xstream';
import { VNode, div, input, span, label } from '@cycle/dom';
import { State } from './index';
import { styles } from './styles';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ description, unit, min, max, step, value }) =>
    div(`.${styles.sliderInput}`, [
      label(description),
      input(`.SliderInput-input .${styles.numberInput}`, {
        attrs: {
          type: 'number',
          min,
          max,
          step
        },
        props: { value }
      }),
      span(`.${styles.sliderInputUnit}`, unit),
      input('.SliderInput-input', {
        attrs: {
          type: 'range',
          min,
          max,
          step
        },
        props: { value }
      })
    ])
  );
}
