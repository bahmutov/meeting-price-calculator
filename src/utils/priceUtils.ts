export const hourlyToTickRatio: number = 1 / 60 / 60;

export default function calculatePrice(
  personAmount: number,
  avgPrice: number,
  tick: number
): number {
  return personAmount * avgPrice * tick * hourlyToTickRatio;
}

export function formatPrice(price: number, currency: string): string {
  // Do not try to format an undefined price
  if (price) {
    const separator = currency === '€' ? ' ' : ',';
    const displayPrice: string = toDisplayPrice(price, separator);
    return currency === '€'
      ? `${displayPrice.replace('.', ',')} ${currency}`
      : `${currency} ${displayPrice}`;
  } else {
    return currency === '€' ? '0,00 €' : '$ 0.00';
  }
}

export function toDisplayPrice(price: number, separator: string): string {
  const roundedPrice = price.toFixed(2);
  const priceParts = roundedPrice.split('.');
  const integerPart: string = priceParts[0];
  const decimalPart: string = priceParts[1];
  const grouped = separateThousands(integerPart, separator);
  return `${grouped}.${decimalPart}`;
}

export function separateThousands(price: string, separator: string): string {
  return price
    .split('')
    .reverse()
    .map((value, index) => {
      if (index !== 0 && index % 3 === 0) {
        return `${value}${separator}`;
      }
      return value;
    })
    .reverse()
    .join('');
}
