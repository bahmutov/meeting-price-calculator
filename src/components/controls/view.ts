import xs from 'xstream';
import * as moment from 'moment';
import { VNode, div, select, option, span, label } from '@cycle/dom';

import { State } from './index';
import { formatPrice } from '../../utils/priceUtils';

function renderOption(currency: string, selectedCurrency: string) {
  return currency === selectedCurrency
    ? option(
        {
          attrs: {
            selected: 'selected'
          }
        },
        [currency]
      )
    : option([currency]);
}

function renderCurrencySelect(selectedCurrency: string) {
  return select('.currency-select', [
    renderOption('â‚¬', selectedCurrency),
    renderOption('$', selectedCurrency)
  ]);
}

export default function view(
  state$: xs<State>,
  personAmountSliderVDom$: xs<VNode>,
  avgPriceSliderVDom$: xs<VNode>
): xs<VNode> {
  return xs
    .combine(state$, personAmountSliderVDom$, avgPriceSliderVDom$)
    .map(
      (
        [{ currency, personAmount, avgPrice }, personAmountVDom, avgPriceVDom]
      ) =>
        div('.Controls', [
          personAmountVDom,
          div('.price-result', [
            label('.total-price-label', 'Total price per hour'),
            div(
              '.total-price-value',
              formatPrice(personAmount.value * avgPrice.value, currency)
            ),
            div('.currency', [
              span('.currency-label.label', 'Currency'),
              renderCurrencySelect(currency)
            ])
          ]),
          avgPriceVDom
        ])
    );
}
