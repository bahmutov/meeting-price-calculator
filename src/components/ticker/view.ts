import xs from 'xstream';
import { VNode, div } from '@cycle/dom';
import { State } from './index';
import { formatPrice } from '../../utils/priceUtils';

export default function view(state$: xs<State>): xs<VNode> {
  return state$.map(({ currency, totalPrice }) =>
    div('.Ticker', [
      div('.Price-actual', [
        div('.Price-label', 'This meeting has cost'),
        div('.Price-value', formatPrice(totalPrice, currency))
      ])
    ])
  );
}
