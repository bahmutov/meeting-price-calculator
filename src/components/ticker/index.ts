import xs from 'xstream';
import * as moment from 'moment';

import { Sources, Sinks } from '../../interfaces';
import { State as AppState } from '../app';
import model from './model';
import view from './view';
import calculatePrice from '../../utils/priceUtils';

export interface State {
  startTime: moment.Moment;
  duration: number;
  currency: string;
  totalPrice: number;
}

export type Reducer = (prev?: State) => State | undefined;

export const lens = {
  get: (state: AppState): State =>
    state
      ? {
          startTime: state.startTime,
          duration: state.duration,
          currency: state.currency,
          totalPrice: calculatePrice(
            state.personAmount,
            state.avgPrice,
            state.duration
          )
        }
      : {
          startTime: moment(),
          duration: 0,
          currency: 'â‚¬',
          totalPrice: 0
        },

  set: (state: AppState, childState: State) => ({
    ...state,
    duration: childState.duration
  })
};

export default function Ticker(sources: Sources): Sinks {
  const reducer$: xs<Reducer> = model(sources.Time);
  const state$: xs<State> = (sources.onion.state$ as any) as xs<State>;
  const vdom$ = view(state$);

  const sinks = {
    DOM: vdom$,
    onion: reducer$
  };

  return sinks;
}
