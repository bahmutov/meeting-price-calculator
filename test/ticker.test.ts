import xs from 'xstream';
import { forall, integer, assert, Options } from 'jsverify';
import { VNode } from '@cycle/dom';
import { withTime } from 'cyclejs-test-helpers';
const htmlLooksLike = require('html-looks-like');
const toHtml = require('snabbdom-to-html');
import * as moment from 'moment';

import { formatPrice } from '../src/utils/priceUtils';
import { State } from '../src/components/ticker';
import view from '../src/components/ticker/view';

const testOptions: Options = {
  tests: 25,
  quiet: false,
  size: 60 * 60 * 24
};

describe('Ticker Component', () => {
  let originalTimeout;

  beforeEach(() => {
    originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;
  });

  afterEach(() => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout;
  });

  it('should match a snapshot correctly', () => {
    const startTime = moment('2017-06-29T21:19:35.849Z');
    const state$: xs<State> = xs.of({
      startTime: startTime,
      duration: 12000,
      currency: '€',
      totalPrice: 123456
    });
    const vdom$: xs<VNode> = view(state$);
    const html$: xs<{}> = vdom$.map(toHtml);
    expect(html$).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const startTime = moment();
    const expectedHTML = (currency: string, totalPrice: number) => {
      const priceSoFar = formatPrice(totalPrice, currency);
      return `
        <div>
          <div>This meeting has cost</div>
          <div>${priceSoFar}</div>
        </div>
      `;
    };

    const property = forall(integer, (tp: number) =>
      withTime(Time => {
        const state$: xs<State> = xs.of({
          startTime: startTime,
          duration: 12000,
          currency: '€',
          totalPrice: tp
        });

        const vdom$ = view(state$);
        const html$ = vdom$.map(toHtml);

        const expected$ = state$.map(({ currency, totalPrice }) =>
          expectedHTML(currency, totalPrice)
        );

        Time.assertEqual(html$, expected$, htmlLooksLike);
      })
    );

    return assert(property, testOptions);
  });
});
