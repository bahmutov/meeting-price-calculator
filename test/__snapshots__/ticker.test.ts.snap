// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticker Component should match a snapshot correctly 1`] = `
Stream {
  "_d": false,
  "_dl": Object {},
  "_err": Object {},
  "_ils": Array [],
  "_prod": MapOp {
    "f": [Function],
    "ins": Stream {
      "_d": false,
      "_dl": Object {},
      "_err": Object {},
      "_ils": Array [],
      "_prod": MapOp {
        "f": [Function],
        "ins": Stream {
          "_d": false,
          "_dl": Object {},
          "_err": Object {},
          "_ils": Array [],
          "_prod": FromArray {
            "a": Array [
              Object {
                "currency": "â‚¬",
                "duration": 12000,
                "startTime": "2017-06-29T21:19:35.849Z",
                "totalPrice": 123456,
              },
            ],
            "type": "fromArray",
          },
          "_stopID": Object {},
          "_target": Object {},
        },
        "out": Object {},
        "type": "map",
      },
      "_stopID": Object {},
      "_target": Object {},
    },
    "out": Object {},
    "type": "map",
  },
  "_stopID": Object {},
  "_target": Object {},
}
`;
